File: golang/langarena/brainfuck_array.go
package LangArena

type Tape struct {
	tape []byte
	pos  int
}

func NewTape() Tape {
	return Tape{tape: make([]byte, 30000), pos: 0}
}

func (t *Tape) Get() byte { return t.tape[t.pos] }

func (t *Tape) Inc() {
	t.tape[t.pos] = t.tape[t.pos] + 1
}

func (t *Tape) Dec() {
	t.tape[t.pos] = t.tape[t.pos] - 1
}

func (t *Tape) Advance() {
	t.pos++
	if t.pos >= len(t.tape) {
		t.tape = append(t.tape, 0)
	}
}

func (t *Tape) Devance() {
	if t.pos > 0 {
		t.pos--
	}
}

type Program struct {
	commands []byte
	jumps    []int
}

func NewProgram(text string) *Program {

	commands := make([]byte, 0, len(text))
	for i := 0; i < len(text); i++ {
		c := text[i]

		switch c {
		case '[', ']', '<', '>', '+', '-', ',', '.':
			commands = append(commands, c)
		}
	}

	jumps := make([]int, len(commands))
	stack := make([]int, 0, len(commands)/2)

	for i, cmd := range commands {
		switch cmd {
		case '[':
			stack = append(stack, i)
		case ']':
			if len(stack) > 0 {
				start := stack[len(stack)-1]
				stack = stack[:len(stack)-1]
				jumps[start] = i
				jumps[i] = start
			}
		}
	}

	return &Program{commands: commands, jumps: jumps}
}

func (p *Program) Run() int64 {
	result := int64(0)
	tape := NewTape()
	pc := 0
	cmds := p.commands
	jumps := p.jumps

	for pc < len(cmds) {
		switch cmds[pc] {
		case '+':
			tape.Inc()
		case '-':
			tape.Dec()
		case '>':
			tape.Advance()
		case '<':
			tape.Devance()
		case '[':
			if tape.Get() == 0 {
				pc = jumps[pc]
				continue
			}
		case ']':
			if tape.Get() != 0 {
				pc = jumps[pc]
				continue
			}
		case '.':
			result = (result << 2) + int64(tape.Get())
		}
		pc++
	}
	return result
}

type BrainfuckArray struct {
	BaseBenchmark
	programText string
	warmupText  string
	result      uint32
}

func (b *BrainfuckArray) Name() string {
	return "Brainfuck::Array"
}

func (b *BrainfuckArray) Prepare() {
	b.programText = b.ConfigStr("program")
	b.warmupText = b.ConfigStr("warmup_program")
	b.result = 0
}

func (b *BrainfuckArray) _Run(text string) int64 {
	return NewProgram(text).Run()
}

func (b *BrainfuckArray) Warmup(bench Benchmark) {
	wi := b.WarmupIterations()
	for i := 0; i < wi; i++ {
		b._Run(b.warmupText)
	}
}

func (b *BrainfuckArray) Run(iteration_id int) {
	runResult := b._Run(b.programText)
	b.result += uint32(runResult)
}

func (b *BrainfuckArray) Checksum() uint32 {
	return b.result
}

File: cpp/src/brainfuck_array.cpp
#include "brainfuck_array.hpp"

BrainfuckArray::Program::Program(const std::string &text) {
  for (char c : text) {
    if (std::string("[]<>+-,.").find(c) != std::string::npos) {
      commands.push_back(static_cast<uint8_t>(c));
    }
  }

  jumps.resize(commands.size(), 0);
  std::vector<size_t> stack;

  for (size_t i = 0; i < commands.size(); ++i) {
    uint8_t cmd = commands[i];
    if (cmd == '[') {
      stack.push_back(i);
    } else if (cmd == ']' && !stack.empty()) {
      size_t start = stack.back();
      stack.pop_back();
      jumps[start] = i;
      jumps[i] = start;
    }
  }
}

int64_t BrainfuckArray::Program::run() {
  int64_t result = 0;
  Tape tape;
  size_t pc = 0;

  while (pc < commands.size()) {
    uint8_t cmd = commands[pc];
    switch (cmd) {
    case '+':
      tape.inc();
      break;
    case '-':
      tape.dec();
      break;
    case '>':
      tape.advance();
      break;
    case '<':
      tape.devance();
      break;
    case '[':
      if (tape.get() == 0) {
        pc = jumps[pc];
        continue;
      }
      break;
    case ']':
      if (tape.get() != 0) {
        pc = jumps[pc];
        continue;
      }
      break;
    case '.':
      result = (result << 2) + static_cast<int64_t>(tape.get());
      break;
    }
    pc++;
  }
  return result;
}

BrainfuckArray::BrainfuckArray() : result_val(0) {
  program_text = Helper::config_s(name(), "program");
  warmup_text = Helper::config_s(name(), "warmup_program");
}

std::string BrainfuckArray::name() const { return "Brainfuck::Array"; }

void BrainfuckArray::warmup() {
  int64_t prepare_iters = warmup_iterations();
  for (int64_t i = 0; i < prepare_iters; i++) {
    _run(warmup_text);
  }
}

void BrainfuckArray::run(int iteration_id) {
  (void)iteration_id;
  int64_t run_result = _run(program_text);
  result_val += static_cast<uint32_t>(run_result);
}

uint32_t BrainfuckArray::checksum() { return result_val; }

int64_t BrainfuckArray::_run(const std::string &text) {
  Program program(text);
  return program.run();
}